parlai/mturk/core/worker_manager.py

```diff
@@ -0,0 +1,344 @@
+
+    def has_assignment(self, assign_id):
+        """Returns true if this worker has an assignment for the given id"""
+        return assign_id in self.agents
+
+    def completed_assignments(self):
+        """Returns the number of assignments this worker has completed"""
+        complete_count = 0
+        for agent in self.agents.values():
+            if agent.get_status() == AssignState.STATUS_DONE:
+                complete_count += 1
+        return complete_count
+
+
+class WorkerManager():
+    """Class used to keep track of workers state, as well as processing
+    messages that come from the web client.
+    """
+
+    def __init__(self, mturk_manager, opt):
+        self.opt = opt
+        self.mturk_manager = mturk_manager
+        self.mturk_workers = {}
+        self.conv_to_agent = {}
+        self.assignment_to_worker_id = {}
+        self.hit_id_to_agent = {}
+        self.time_blocked_workers = []
+        self.load_disconnects()
+        self.is_sandbox = mturk_manager.is_sandbox
+
+    def _create_agent(self, hit_id, assignment_id, worker_id):
+        """Initialize an agent and return it"""
+        return MTurkAgent(
+            self.opt, self.mturk_manager, hit_id, assignment_id, worker_id)
+
+    def _get_worker(self, worker_id):
+        """A safe way to get a worker by worker_id"""
+        if worker_id in self.mturk_workers:
+            return self.mturk_workers[worker_id]
+        return None
+
+    def _get_agent(self, worker_id, assignment_id):
+        """A safe way to get an agent by worker and assignment_id"""
+        worker = self._get_worker(worker_id)
+        if worker is not None:
+            if assignment_id in worker.agents:
+                return worker.agents[assignment_id]
+        return None
+
+    def route_packet(self, pkt):
+        """Put an incoming message into the queue for the agent specified in
+        the packet, as they have sent a message from the web client.
+        """
+        worker_id = pkt.sender_id
+        assignment_id = pkt.assignment_id
+        agent = self._get_agent(worker_id, assignment_id)
+        if agent is not None:
+            shared_utils.print_and_log(
+                logging.INFO,
+                'Manager received: {}'.format(pkt),
+                should_print=self.opt['verbose']
+            )
+            # Push the message to the message thread to send on a reconnect
+            agent.append_message(pkt.data)
+
+            # Clear the send message command, as a message was recieved
+            agent.set_last_command(None)
+            agent.put_data(pkt.id, pkt.data)
+
+    def map_over_agents(self, map_function, filter_func=None):
+        """Take an action over all the agents we have access to, filters if
+        a filter_func is given"""
+        for worker in self.mturk_workers.values():
+            for agent in worker.agents.values():
+                if filter is None or filter_func(agent):
```

Comment:
i think `filter` should be `filter_func` here

Reply:
Good catch! Turns out I started failing this test. Oops...